include/zephyr/net/dhcpv4_server.h

```diff
@@ -107,6 +107,35 @@ int net_dhcpv4_server_foreach_lease(struct net_if *iface,
 				    net_dhcpv4_lease_cb_t cb,
 				    void *user_data);
 
+/**
+ * @typedef net_dhcpv4_server_provider_cb_t
+ * @brief Callback used to let application provide an address for a given
+ * client ID
+ * @details This function is called before assigning an address to a client,
+ * and lets the application override the address for a given client. If the
+ * callback returns 0, addr needs to be a valid address and will be assigned
+ * to the client. If the callback returns anything non-zero, the client will
+ * be assigned an address from the pool.
+ *
+ * @param iface Pointer to the network interface
+ * @param client_id Pointer to client requesting an address
+ * @param addr Address to be assigned to client
+ * @param user_data A valid pointer to user data or NULL
+ */
+typedef int (*net_dhcpv4_server_provider_cb_t)(struct net_if *iface,
+											   const struct dhcpv4_client_id *client_id,
```

Comment:
Something went wrong here and elsewhere, note that Zephyr uses 8 spaces/tab for indentation.